src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,297 @@
+                ImmutableArray<(IParameterSymbol, ImmutableArray<SyntaxNode>)> parametersNeedGenerateDeclarations,
+                ImmutableArray<IParameterSymbol> parametersWithDefaultValue,
+                ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode literalExpressionSyntaxNode)> parametersWithLiteralArgument)
+            {
+                ParametersWithIdentifierArgument = parametersWithIdentifierArgument;
+                ParametersWithVariableDeclarationArgument = parametersWithVariableDeclarationArgument;
+                ParametersNeedGenerateDeclarations = parametersNeedGenerateDeclarations;
+                ParametersWithDefaultValue = parametersWithDefaultValue;
+                ParametersWithLiteralArgument = parametersWithLiteralArgument;
+            }
+
+            public static MethodParametersInfo GetMethodParametersInfo(
+                AbstractInlineMethodRefactoringProvider<TInvocationSyntaxNode, TExpressionSyntax, TArgumentSyntax> inlineMethodRefactoringProvider,
+                ISyntaxFacts syntaxFacts,
+                SemanticModel semanticModel,
+                SyntaxNode calleeInvocationSyntaxNode,
+                IMethodSymbol calleeMethodSymbol,
+                CancellationToken cancellationToken)
+            {
+                // Classify the parameters into:
+                // 1. If the parameter accept an identifier from Caller, then use this identifier to replace all the occurence of the parameter.
+                // (So that after inlining, the same identifier could be found in the caller)
+                // 2. If the parameter accept literal/no argument but has default value, replace all the occurence of the parameter by using the
+                // literal expression
+                // 3. If the argument has variable declarations ('out var' in C#), then insert an additional declaration. Also replace all the
+                // occurence of parameter by using this identifier.
+                // 4. For the rest of the parameters (might include method invocation and different expressions), use the parameter's name to
+                // generate a declaration for them and insert that into caller.
+                var allArguments = syntaxFacts.GetArgumentsOfInvocationExpression(calleeInvocationSyntaxNode);
+                var parameterSymbolAndArguments = allArguments
+                    .Select(argument => (
+                        parameterSymbol: inlineMethodRefactoringProvider.GetParameterSymbol(semanticModel, (TArgumentSyntax)argument,
+                            cancellationToken),
+                        argumentExpression: syntaxFacts.GetExpressionOfArgument(argument)))
+                    .Where(parameterAndArgument =>
+                        parameterAndArgument.parameterSymbol != null
+                        && !parameterAndArgument.parameterSymbol.IsDiscard)
```

Comment:
1. needs docs on why you're doing the IsDicard check.
2. please don't put the comments inline in these big expressions.  that conveys that things are too complex.  If you're at the point you need to comment, just break into simple statements with easy to follow steps.  i.e. massive single expressions are not a virtue.

Reply:
I removed this part when i try to use the IOperation to handle the extension method case. :) Actually it make this part of code a lot simpler so I will mark this as resolved.